text/0000-builder-spec.md

```diff
@@ -0,0 +1,295 @@
+  samples/hello-moon            0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/hello-moon
+
+Detection Order:
+ ├ Group #1:
+ │  └ samples/java-maven@0.0.1
+ ├ Group #2:
+ │  └ samples/kotlin-gradle@0.0.1
+ ├ Group #3:
+ │  └ samples/ruby-bundler@0.0.1
+ └ Group #4:
+    └ samples/hello-universe@0.0.1
+       └ Group #1:
+          ├ samples/hello-world@0.0.1
+          └ samples/hello-moon@0.0.1
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+## Structure of Builders
+### Files/Directories
+A builder is composed of at least the following directories/files:
+```
+- /cnb/buildpacks/...<buildpack ID>/<buildpack version>/
+- /cnb/lifecycle/<lifecycle binaries>
+- /cnb/order.toml
+- /cnb/stack.toml
+- /layers
+- /workspace
+```
+
+In addition, builders MAY build in environment variables by defining env-files in
+```
+- /platform/env/<env-files>
+```
+
+### Env Vars/Labels
+A builder's environment is the build-time environment of the stack, and as such, requires all of the [build image specifications][build-image-specs].
+
+Additionally, a builder requires:
+* The image config's WorkingDir should be set
+* The image config's Label field has the label io.buildpacks.buildpack.order, set to a JSON object representing an [Order](#order)
+* The image config's Label field has the label io.buildpacks.builder.metadata, set to a JSON object representing [Builder Metadata](#metadata)
+* The image config's Label field has the label io.buildpacks.buildpack.layers, set to a JSON object representing the [layers](#layers)
+
+If the builder contains an optional [lifecycle descriptor file][lifecycle-descriptor-rfc], it also requires:
+* The image config's Label field has the label io.buildpacks.lifecycle.version, set to the lifecycle version
+* The image config's Label field has the label io.buildpacks.lifecycle.apis, set to a JSON object representing the [lifecycle apis](#lifecycle-apis)
+
+#### Order
+The `io.buildpacks.buildpack.order` data should look like:
+```json
+[
+	{
+	  "group":
+		[
+		  {
+			"id": "<buildpack id>",
+			"version": "<buildpack version>",
+			"optional": "<bool>"
+		  }
+		]
+	}
+]
+```
+
+#### Metadata
+The `io.buildpacks.builder.metadata` data should look like:
+```json
+{
+  "description": "<Some description>",
+  "api": "<some builder API version | default 0.1 >",
+  "stack": {
+    "runImage": {
+      "image": "<some/run-image>",
+      "mirrors": [
+        "<gcr.io/some/default>"
+      ]
+    }
+  },
+  "buildpacks": [
+    {
+      "id": "<buildpack id>",
+	  "version": "<buildpack version>",
+	  "homepage": "http://geocities.com/top-bp"
+	}
+  ],
+  "lifecycle": {
+    "version": "<version of lifecycle in builder>",
+    "apis":  {
+        "buildpack": {
+          "deprecated": ["0.1"],
+          "supported": ["1.2", "1.3"]},
+        "platform": {
+          "deprecated": [],
+          "supported": ["2.3", "2.4"]}
+    }
+  },
```

Comment:
Given this information exists in the `io.buildpacks.lifecycle.*` labels I think we can pull it out of the builder metadata. The only reason the lifecycle version and the old form of the APIs:
```
  "lifecycle": {
    "version": "0.9.1",
    "api": {
      "buildpack": "0.2",
      "platform": "0.3"
      }
   },
   ...
```
are in the current label is for backwards compatibility. I don't think we should specifying the backwards compatible data, even if we continue to include in `pack create-builder` for some period of time.